wp-includes/class-wp-service-workers.php

```diff
@@ -0,0 +1,221 @@
+	 * @param array           $scopes Optional. Scopes of the service worker. Default relative path.
+	 * @return bool Whether the item has been registered. True on success, false on failure.
+	 */
+	public function register( $handle, $src, $deps = array(), $scopes = array() ) {
+
+		// Set default scope if missing.
+		if ( empty( $scopes ) ) {
+			$scopes = array( site_url( '/', 'relative' ) );
+		}
+		return parent::add( $handle, $src, $deps, false, compact( 'scopes' ) );
+	}
+
+	/**
+	 * Get service worker logic for scope.
+	 *
+	 * @param string $scope Scope of the Service Worker.
+	 */
+	public function serve_request( $scope ) {
+
+		header( 'Content-Type: text/javascript; charset=utf-8' );
+		nocache_headers();
+
+		if ( ! in_array( $scope, $this->get_scopes(), true ) ) {
+			status_header( 404 );
+			return;
+		}
+
+		$scope_items = array();
+
+		// Get handles from the relevant scope only.
+		foreach ( $this->registered as $handle => $item ) {
+			if ( in_array( $scope, $item->args['scopes'], true ) ) {
+				$scope_items[] = $handle;
+			}
+		}
+
+		$this->output = '';
+		$this->do_items( $scope_items );
+
+		$file_hash = md5( $this->output );
+		header( "Etag: $file_hash" );
+
+		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
+		if ( $file_hash === $etag_header ) {
+			status_header( 304 );
+			return;
+		}
+
+		// @codingStandardsIgnoreLine
+		echo $this->output;
+	}
+
+	/**
+	 * Get all scopes.
+	 *
+	 * @return array Array of scopes.
+	 */
+	public function get_scopes() {
+
+		$scopes = array();
+		foreach ( $this->registered as $handle => $item ) {
+			$scopes = array_merge( $scopes, $item->args['scopes'] );
+		}
+		return array_unique( $scopes );
+	}
+
+	/**
+	 * Process one registered script.
+	 *
+	 * @param string $handle Handle.
+	 * @param bool   $group Group.
+	 * @return void
+	 */
+	public function do_item( $handle, $group = false ) {
+		global $wp_filesystem;
+
+		if ( null === $wp_filesystem ) {
+			WP_Filesystem();
+		}
+
+		$obj = $this->registered[ $handle ];
+
+		if ( is_callable( $obj->src ) ) {
+			$this->output .= call_user_func( $obj->src ) . "\n";
+		} else {
+			$validated_path = $this->get_validated_file_path( $obj->src );
+			if ( is_wp_error( $validated_path ) ) {
+				_doing_it_wrong(
+					__FUNCTION__,
+					/* translators: %s is file URL */
+					sprintf( esc_html__( 'Service worker src is incorrect: %s', 'pwa' ), esc_html( $obj->src ) ),
+					'0.1'
+				);
+
+				/* translators: %s is file URL */
+				$this->output .= "console.warn( '" . sprintf( esc_html__( 'Service worker src is incorrect: %s', 'pwa' ), esc_html( $obj->src ) ) . "' );\n";
+			} else {
+				/* translators: %s is file URL */
+				$this->output .= sprintf( esc_html( "\n/* Source: %s */\n" ), esc_url( $obj->src ) );
+				$this->output .= $wp_filesystem->get_contents( $this->get_validated_file_path( $obj->src ) ) . "\n";
```

Comment:
I wonder actually if we need to use `$wp_filesystem->get_contents()` and shouldn't just use `file_get_contents()`? Core actually just uses `file_get_contents()` when it needs to get a file from the filesystem, for example:

https://github.com/WordPress/wordpress-develop/blob/97cacadc50716695d8c8d3afe129d901f4230266/src/wp-includes/class-wp-theme.php#L1133-L1138

I don't see any core uses of calls to `$wp_filesystem->get_contents()` at all.

Reply:
Saw that as well, wasn't sure if it's because the `@file_get_contents` was in use already before the `WP_Filesystem` or if it's just OK for core.

If that's OK within core then we can just remove using the `WP_Filesystem` and add ignore for the coding standards.

Reply:
Yes, let's do that. The only reason for the coding standards warning is the usage of this function to make HTTP requests. Since we are validating the path is local to the filesystem, the coding standards warning does not apply.